packages/snaps-controllers/src/services/webview/WebViewMessageStream.ts

```diff
@@ -0,0 +1,97 @@
+import { logError } from '@metamask/snaps-utils';
+import { assert, bytesToBase64, stringToBytes } from '@metamask/utils';
+
+export type WebViewInterface = {
+  injectJavaScript(js: string): void;
+  registerMessageListener(listener: (event: MessageEvent) => void): void;
+  unregisterMessageListener(listener: (event: MessageEvent) => void): void;
+};
+
+type WebViewStreamArgs = {
+  name: string;
+  target: string;
+  getWebView: () => Promise<WebViewInterface>;
+};
+
+/**
+ * A special postMessage stream used to interface with a WebView.
+ */
+
+export class WebViewMessageStream extends BasePostMessageStream {
+  #name;
+
+  #target;
+
+  #webView: WebViewInterface | undefined;
+
+  /**
+   * Creates a stream for communicating with other streams inside a WebView.
+   *
+   * @param args - Options bag.
+   * @param args.name - The name of the stream. Used to differentiate between
+   * multiple streams sharing the same window object.
+   * @param args.target - The name of the stream to exchange messages with.
+   * @param args.getWebView - A asynchronous getter for the webview.
+   */
+  constructor({ name, target, getWebView }: WebViewStreamArgs) {
+    super();
+
+    this.#name = name;
+    this.#target = target;
+
+    this._onMessage = this._onMessage.bind(this);
+
+    // This is a bit atypical from other post-message streams.
+    // We have to wait for the WebView to fully load before we can continue using the stream.
+    getWebView()
+      .then((webView) => {
+        this.#webView = webView;
+        webView.registerMessageListener(
+          (event: MessageEvent<PostMessageEvent>) =>
+            this._onMessage(event as any),
+        );
+        this._handshake();
+      })
+      .catch((error) => {
+        logError(error);
+      });
+  }
+
+  protected _postMessage(data: unknown): void {
+    assert(this.#webView);
+    const json = JSON.stringify({
+      target: this.#target,
+      data,
+    });
+
+    // To prevent XSS, we base64 encode the message before injecting it.
+    // This adds significant performance overhead.
+    // TODO: Should we use mobile native base64 here?
+    const bytes = stringToBytes(json);
+    const base64 = bytesToBase64(bytes);
+    this.#webView.injectJavaScript(`window.postMessage('${base64}')`);
+  }
+
+  private _onMessage(event: PostMessageEvent): void {
+    const message = JSON.parse(event.data as string);
+
+    // Notice that we don't check targetWindow or targetOrigin here.
+    // This doesn't seem possible to do in RN.
+    // TODO: Review whether we are fine with this before using in production.
+    if (!isValidStreamMessage(message) || message.target !== this.#name) {
+      return;
+    }
+
+    this._onData(message.data);
+  }
+
+  _destroy() {
+    assert(this.#webView);
+    this.#webView.unregisterMessageListener(
+      (event: MessageEvent<PostMessageEvent>) => this._onMessage(event as any),
```

Comment:
```suggestion
      this._onMessage,
```
Could we pass this directly?